src/compiler/parse/read/context.ts

```diff
@@ -1,202 +1,72 @@
-	}
-
-	else if (parser.eat('{')) {
-		context.type = 'ObjectPattern';
-		context.properties = [];
-
-		do {
-			parser.allow_whitespace();
-
-			if (parser.eat('...')) {
-				parser.allow_whitespace();
-
-				const start = parser.index;
-				const name = parser.read_identifier();
-				const key: Identifier = {
-					start,
-					end: parser.index,
-					type: 'Identifier',
-					name
-				};
-				const property: Property = {
-					start,
-					end: parser.index,
-					type: 'Property',
-					kind: 'rest',
-					shorthand: true,
-					key,
-					value: key
-				};
-
-				context.properties.push(property);
-
-				parser.allow_whitespace();
-
-				if (parser.eat(',')) {
-					parser.error({
-						code: `comma-after-rest`,
-						message: `Comma is not permitted after the rest element`
-					}, parser.index - 1);
-				}
-
+			bracket_stack.pop();
+			if (bracket_stack.length === 0) {
+				i += code <= 0xffff ? 1 : 2;
 				break;
 			}
-
-			// TODO: DRY this out somehow
-			// We don't know whether we want to allow reserved words until we see whether there's a ':' after it
-			// Probably ideally we'd use Acorn to do all of this
-			const start = parser.index;
-			const name = parser.read_identifier(true);
-			const key: Identifier = {
-				start,
-				end: parser.index,
-				type: 'Identifier',
-				name
-			};
-			parser.allow_whitespace();
-
-			let value: Context;
-			if (parser.eat(':')) {
-				parser.allow_whitespace();
-				value = read_context(parser);
-			} else {
-				if (reserved.has(name)) {
-					parser.error({
-						code: `unexpected-reserved-word`,
-						message: `'${name}' is a reserved word in JavaScript and cannot be used here`
-					}, start);
-				}
-				value = key;
-			}
-
-			const property: Property = {
-				start,
-				end: value.end,
-				type: 'Property',
-				kind: 'init',
-				shorthand: value.type === 'Identifier' && value.name === name,
-				key,
-				value
-			};
-
-			context.properties.push(property);
-
-			parser.allow_whitespace();
-		} while (parser.eat(','));
-
-		error_on_assignment_pattern(parser);
-		parser.eat('}', true);
-		context.end = parser.index;
-	}
-
-	else if (parser.eat('...')) {
-		const name = parser.read_identifier();
-		if (name) {
-			context.type = 'RestIdentifier';
-			context.end = parser.index;
-			context.name = name;
-		}
-
-		else {
-			parser.error({
-				code: 'invalid-context',
-				message: 'Expected a rest pattern'
-			});
 		}
+		i += code <= 0xffff ? 1 : 2;
 	}
 
-	else {
-		const name = parser.read_identifier();
-		if (name) {
-			context.type = 'Identifier';
-			context.end = parser.index;
-			context.name = name;
-		}
+	parser.index = i;
 
-		else {
-			parser.error({
-				code: 'invalid-context',
-				message: 'Expected a name, array pattern or object pattern'
-			});
-		}
-
-		error_on_assignment_pattern(parser);
+	const pattern_string = parser.template.slice(start, i);
+	try {
+		return (parse_expression_at(
+			`${" ".repeat(start - 1)}(${pattern_string} = 1)`,
```

Comment:
Is this going to cause a problem with sourcemap alignment and newlines? Should we use something like cf72b4157575e4a64d73c75dc26fa22bcd25f719?

Reply:
fixed and updated the test case to test this too.